docs/source/user_guides/basic_conceptions.rst

```diff
@@ -139,165 +148,185 @@ That's what the agent consists of:
+If you retrieve data from the database or API, it's important to validate it to ensure it meets expectations. 
+Since DFF extensively leverages pydantic, you can resort to the validation tools of this feature-rich library.
+For instance, given that each processing routine is a callback, you can use tools like pydantic's `validate_call`
+to ensure that the returned values match the function signature.
+Error handling logic can also be incorporated into the callbacks.
+
+Generating a bot Response
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Generating a bot response involves creating a text or multimedia response that will be delivered to the user.
+Response is defined in the ``RESPONSE`` section of each node and should be either a ``Message`` object,
+that can contain text, images, audios, attachments, etc., or a callback that returns a ``Message``.
+The latter allows you to customize the response based on the specific scenario and user input.
+
+.. code-block:: python
+
+    def sample_response(ctx: Context, _: Pipeline, *args, **kwargs) -> Message:
+        if ctx.misc["user"] == 'vegan':
+            return Message(text="Here is a list of vegan cafes.")
+        return Message(text="Here is a list of cafes.")
+
+Handling Fallbacks
+~~~~~~~~~~~~~~~~~~
+
+In DFF, you should provide handling for situations where the user makes requests
+that do not trigger any of the transitions specified in the script graph. 
+To cover that use case, DFF requires you to define a fallback node that the agent will move to
+when no adequate transition has been found.
+
+Like other nodes, the fallback node can either use a message or a callback to produce a response
+which gives you a lot of freedom in creating situationally appropriate error messages.
+Create friendly error messages and, if possible, suggest alternative options. 
+This ensures a smoother user experience even when the bot encounters unexpected inputs.
+
+.. code-block:: python
 
     def fallback_response(ctx: Context, _: Pipeline, *args, **kwargs) -> Message:
         """
-        Generate response for fallback node, according to the previous node
-        we have been to.
-        If the previous node was `start_node`, a sample message will be returned,
-        otherwise the message will include user input.
+        Generate a special fallback response if the initial user utterance is not 'Hi'.
         """
-        if get_previous_node_name(ctx) == "start_node":
+        last_flow, last_node = ctx.last_label
+        if ctx.last_request.text != "Hi" and last_node == "start_node":
             return Message(text="You should've started the dialog with 'Hello!'")
-        elif ctx.last_request is not None:
-            last_request = ctx.last_request.text
-            note = f"You should've written 'Ping', not '{last_request}'!"
-            return Message(text=f"That was against the rules! {note}")
         else:
             raise RuntimeError("Error occurred: last request is None!")
```

Comment:
This function doesn't make sense now.:
1. "Hi" vs "Hello!"; 
2. An error is raised even if `last_request` is not None (e.g. if `last_node` is not "start_node").